feat: Add CRUD operations for notes and categories

sherif566 opened the pull request:
**Phase 1: CRUD:**

1. Implement CRUD operations for notes and categories.
2. Use Request classes for validation.
3. Use Response classes for JSON responses.
4. Implement Repositories and Services.
5. Ensured controllers contain no logic.

sherif566 committed:
feat: add CRUD operations for notes and categories

sherif566 committed:
feat: Add repositories to the services and controllers

sherif566 committed:
chore: Add request classes for validation

NadaMohamedOptimalSolutions commented on how 190ecbb:
Please remove this file

NadaMohamedOptimalSolutions commented on routes/api.php:
```diff
@@ -17,3 +19,8 @@
 });
+
+
+Route::resource('categories', CategoryController::class);
```
Please use Route::apiResource instead of Route::resource to create only the REST API CRUD routes. As Route::resource generates routes for both web-based views (like create and edit methods) and API actions. Since this application is API-only, those extra routes are unnecessary. Route::apiResource avoids creating these view-related routes, keeping the API clean and focused on essential CRUD operations.

NadaMohamedOptimalSolutions commented on database/migrations/2024_12_12_124701_create_categories_table.php:
```diff
@@ -0,0 +1,31 @@
+    public function up(): void
+    {
+        Schema::create('categories', function (Blueprint $table) {
+            $table->engine = 'InnoDB'; // Ensure InnoDB engine is used
```
Please remove this line. In the database config files the default engine will be specified.

NadaMohamedOptimalSolutions commented on database/seeders/CategorySeeder.php:
```diff
@@ -0,0 +1,15 @@
+{
+    public function run(): void
+    {
+        Category::factory()->count(10)->create();
```
Please use Category::factory(10)->create() instead of Category::factory()->count(10)->create(), as this is the convention we use. It is also more concise since the factory method accepts a number directly to specify the number of instances to create.

NadaMohamedOptimalSolutions commented on database/factories/CategoryFactory.php:
```diff
@@ -0,0 +1,20 @@
+    {
+        return [
+            'name' => $this->faker->word,
+            'parent_id' => null
```
Please remove parent_id as it has a default value of null already in the database

NadaMohamedOptimalSolutions commented on database/factories/NoteFactory.php:
```diff
@@ -0,0 +1,22 @@
+        return [
+            'title' => $this->faker->sentence,
+            'description' => $this->faker->paragraph,
+            'category_id' => Category::inRandomOrder()->first()->id
```
Please use Category::factory()->create()->id, to make sure that a new instance of the model is created. As Category::inRandomOrder() does not create a category. It only queries the existing records in the categories table and randomizes their order. If we created a note with the an empty categories table category_id will not have a value.

NadaMohamedOptimalSolutions commented on app/Models/Category.php:
```diff
@@ -0,0 +1,29 @@
+
+    protected $fillable = ['name', 'parent_id'];
+
+    // A category can have many notes
```
Please remove the comments, as they do not add any new information.

NadaMohamedOptimalSolutions commented on app/Http/Requests/UpdateCategoryRequest.php:
As store, update request have the same rules you can combine them in a single request class. Please make sure that the new class name is clear.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,54 @@
+    protected $categoryService;
+
+    // Constructor to inject CategoryService
+    public function __construct(CategoryService $categoryService)
```
You can simplify the constructor by using __construct(private CategoryService $categoryService). This approach eliminates the need to define the $categoryService property and assign it manually in the constructor, making the code cleaner and more concise.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    // Get a category by ID
+    public function show($id)
```
Please use model route binding. Model route binding allows Laravel to automatically resolve the model instance from the route parameter, making the code cleaner and more readable. It also handles 404 errors automatically if the model is not found, improving the application's reliability and user experience.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,54 @@
+
+    // Create a new category
+
+    public function store(StoreCategoryRequest $request)
```
Please use a DTO to send data instead of sending an array. Using a DTO makes it clear what data is being sent and enforces the types of the data, leading to better type safety, readability, and maintainability of the code.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    // Get all categories
+    public function index()
```
Could you please add a way to search the categories?

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,54 @@
+    public function index()
+    {
+        $categories = $this->categoryService->getAllCategories();
+        return response()->json($categories);
```
Please define a resource class to return the data. This ensures a consistent and structured format for API responses. Also, ensure that the index results are paginated to handle large datasets efficiently. Each endpoint in the application should have a unified response structure to maintain consistency and improve the API's readability and usability.

NadaMohamedOptimalSolutions commented on app/Services/CategoryService.php:
```diff
@@ -0,0 +1,49 @@
+        $category = $this->categoryRepository->findById($id);
+
+        // Check if the note exists
+        if (!$category) {
```
How will this exception be handled? What will be returned in the response?

NadaMohamedOptimalSolutions commented on app/Services/CategoryService.php:
```diff
@@ -0,0 +1,49 @@
+        $this->categoryRepository = $categoryRepository;
+    }
+
+    public function getAllCategories()
```
You can remove the word category from the method names, as they are all under the CategoryService. Repeating "category" in the method names does not add any new information.

NadaMohamedOptimalSolutions commented on app/Services/CategoryService.php:
```diff
@@ -0,0 +1,49 @@
+        return $this->categoryRepository->getAll();
+    }
+
+    public function findCategoryById($id)
```
Please define return type of for the methods, to increase the code readability.

NadaMohamedOptimalSolutions commented on app/Repositories/CategoryRepository.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function update(Category $category, array $data)
+    {
+        $category->update($data);
```
Please use tap method to return the updated model.

NadaMohamedOptimalSolutions commented on app/Repositories/CategoryRepository.php:
```diff
@@ -0,0 +1,35 @@
+
+    public function delete(Category $category)
+    {
+        $category->delete();
```
You can return $category->delete(); directly, as delete() returns boolean.

NadaMohamedOptimalSolutions reviewed (changes requested):
If the comment is applicable somewhere else please also resolve it. I will not repeat the comments as they will add no value just noise. 

1. Why is the main branch empty? Where is Laravel project structure, config files and so on? 
2. Commits should be separated, so please next time make sure they are
3. Please add a nested endpoint to manage notes through categories 
4. Error handling, for example if model not found exception is throw what will be return in the response? 
5. Please add logs to the application. Make sure to add the appropriate log type.

sherif566 committed:
feat: added a nested endpoint to manage notes through categories

sherif566 committed:
feat: added exception handling and adjusted response jsons to include a message

sherif566 committed:
feat: return all notes under a certain category

NadaMohamedOptimalSolutions commented on app/Repositories/CategoryRepository.php:
Please make sure that this repository implements an interface. Make sure that the interface is reusable.

sherif566 committed:
Add new features for category-specific note operations

sherif566 committed:
Add LogRequest middleware to handle request logging globally

sherif566 committed:
feat: Add AccessLog middleware for logging incoming requests

sherif566 committed:
refactor: Combine Store and Update request classes into single NoteRequest and CategoryRequest

sherif566 committed:
refactor: Implement model route binding, relationship validation, HTTP status standardization, and critical action logging

sherif566 committed:
Added logging to CategoryController methods and updated repository logic for pagination and added DTO's and included usage of tap function

sherif566 committed:
Implement repository interfaces for category and note management

sherif566 committed:
Adjust code based on PR review comments

NadaMohamedOptimalSolutions commented on routes/api.php:
```diff
@@ -2,18 +2,29 @@
+
+Route::apiResource('notes', NoteController::class);
+
+Route::prefix('categories')->group(function () {
```
Please use apiResource to define the nested route instead of creating every route manually.

NadaMohamedOptimalSolutions commented on routes/web.php:
Please remove this file, as we do not need it.

NadaMohamedOptimalSolutions commented on app/Services/NoteService.php:
```diff
@@ -0,0 +1,81 @@
+
+    public function updateNoteInCategory(Note $note,NoteDTO $dto, Category $category)
+    {
+        if ($note->category_id !== $category->id) {
```
Think of a better approach to handle this validation as it is repeated

NadaMohamedOptimalSolutions commented on app/Repositories/NoteRepository.php:
Both note and category interfaces share methods, how can we refactor them?

NadaMohamedOptimalSolutions commented on app/Providers/AppServiceProvider.php:
Do we need to define the bindings here? Could they be automatically resolved?

NadaMohamedOptimalSolutions commented on app/Interfaces/NoteRepositoryInterface.php:
Remove this class as it is empty.

NadaMohamedOptimalSolutions commented on app/Interfaces/RepositoryInterface.php:
Please add the Interfaces directory under Repositories directory

NadaMohamedOptimalSolutions commented on app/Http/Resources/NoteResource.php:
```diff
@@ -0,0 +1,26 @@
+            'id' => $this->id,
+            'title' => $this->title,
+            'description' => $this->description,
+            'category_id' => $this->category_id,
```
Please return the category data here not just the id.

NadaMohamedOptimalSolutions replied:
Please remove `category_id` from the response

NadaMohamedOptimalSolutions commented on app/Http/Resources/CategoryResource.php:
```diff
@@ -0,0 +1,25 @@
+            'name' => $this->name,
+            'created_at' => $this->created_at,
+            'updated_at' => $this->updated_at,
+            'notes' => NoteResource::collection($this->whenLoaded('notes'))
```
What will happen if the notes relation is not loaded?

NadaMohamedOptimalSolutions commented on app/Http/Requests/CategoryRequest.php:
```diff
@@ -0,0 +1,21 @@
+    {
+        return [
+            'name' => 'required|string|max:255',
+            'parent_id' => 'nullable|exists:categories,id',
```
Please use Category::class instead of the categories table name to avoid hardcoding database table names in the code. This approach improves maintainability, as the table name can be defined or changed in the model itself if needed, without affecting the rest of the code.

NadaMohamedOptimalSolutions commented on routes/api.php:
```diff
@@ -2,18 +2,29 @@
-|
 */
 
 Route::middleware('auth:sanctum')->get('/user', function (Request $request) {
```
Please remove this route as it is not used

NadaMohamedOptimalSolutions reviewed (changes requested):
Please add methods return types.

We will meet inshAllah and discuss some comments in detail

sherif566 committed:
Add CategoryNoteController &  CategoryNoteService to handle category-specific note operations

sherif566 committed:
Refactor directory structure to move interfaces under Repositories

sherif566 committed:
Enhance NoteResource to include detailed category data

sherif566 committed:
Remove unused '/user' route from API routes

sherif566 committed:
Use dynamic table name retrieval in validation rules

sherif566 committed:
Add return types to Controller  methods

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,64 @@
+
+    public function store(CategoryRequest $request): CategoryResource
+    {
+        $dto = new CategoryDTO(
```
You can use CategoryDTO::from($request->validated()) instead. This ensures that only the validated request data is used, and most importantly, when a new field is added to the request, you won't need to manually add it here. It will automatically be returned from the validated() method.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,91 @@
+
+    public function index($category): AnonymousResourceCollection
+    {
+        $notes = Note::where('category_id', $category)->paginate(10);
```
Please use the service to get the notes, avoid adding any logic in the controller.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,91 @@
+
+    public function destroy(Category $category, Note $note): JsonResponse
+    {
+        if ($note->category_id !== $category->id) {
```
This validation is repeated. To avoid duplicating the code, you can create a validator class with a method to validate that a note belongs to a category. Then, simply call this method wherever you need to perform the validation. This will improve code reusability and maintainability.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/NoteController.php:
```diff
@@ -0,0 +1,66 @@
+    {
+    }
+
+    public function index(): AnonymousResourceCollection
```
Please make sure that your logging is consistent, in CategoryController you added logs in most methods, however here you did not add the logs.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/NoteController.php:
```diff
@@ -0,0 +1,66 @@
+    {
+        Log::info("Deleting note", ['note_id' => $note->id]);
+        $this->noteService->delete($note);
+        Log::info("Note deleted successfully", ['note_id' => $note->id]);
```
Please do not mention the note ID in this log, as the record is deleted.

NadaMohamedOptimalSolutions commented on app/Http/Requests/CategoryRequest.php:
```diff
@@ -0,0 +1,22 @@
+    {
+        return [
+            'name' => 'required|string|max:255',
+            'parent_id' => 'nullable|exists:' . Category::getTableName() . ',id',
```
You can use Category::class, so 'nullable|exists:' . Category::class . ',id', and Laravel will handle retrieving the model's table name. There's no need to add a new getTableName method to get the table name.

NadaMohamedOptimalSolutions commented on app/Models/Category.php:
```diff
@@ -0,0 +1,30 @@
+    {
+        return (new self())->getTable();
+    }
+    use HasFactory;
```
Please make sure the the class is formatted correctly. First the uses, then the methods. So add use HasFactory; before getTableName method

NadaMohamedOptimalSolutions commented on app/Repositories/Interfaces/RepositoryInterface.php:
```diff
@@ -0,0 +1,11 @@
+
+interface RepositoryInterface
+{
+    public function all($perPage = 10);
```
Instead of using the value 10, you can create a class to handle the pagination number and set the pagination value there. This way, if you need to change the pagination value, you can do it in a single file, ensuring consistency across the application. This class canhandle the constant values, you can define pagination as a const in it.

NadaMohamedOptimalSolutions commented on app/Repositories/Interfaces/RepositoryInterface.php:
```diff
@@ -0,0 +1,11 @@
+
+namespace App\Repositories\Interfaces;
+
+interface RepositoryInterface
```
Amazing job on the repository!
Please name it to something more clear. What I mean is that this repository handles only the CRUD operations, so what can we name it?

NadaMohamedOptimalSolutions commented on app/Repositories/NoteRepository.php:
```diff
@@ -0,0 +1,40 @@
+
+    public function create(array $data)
+    {
+        return Note::create($data);
```
The creation functionality is repeated between the note and category repositories, how can we handle this more efficiently so we do not need to repeat the code?

NadaMohamedOptimalSolutions commented on app/Services/CategoryNoteService.php:
```diff
@@ -0,0 +1,50 @@
+
+    public function createNoteInCategory(NoteDTO $dto, Category $category)
+    {
+        return $this->noteRepository->create([
```
You can use spaite Laravel data package to create the DTOs, as this package provider many helpful method, for example toArray() method, which will turn the DTO into an array and you won't need to define the casting.

NadaMohamedOptimalSolutions commented on app/Services/CategoryNoteService.php:
```diff
@@ -0,0 +1,50 @@
+        ]);
+    }
+
+    public function updateNoteInCategory(Note $note, NoteDTO $dto, Category $category)
```
You can name this method update, as you are in the CategoryNoteService, and the context already makes it clear what the method refers to. Adding "NoteInCategory" is redundant.

NadaMohamedOptimalSolutions reviewed (changes requested):
1. Please insure consistency while logging, as some controller have logs other don’t. Some methods have logs, other don’t
2. Make sure that the commits follow the guide defined in the task requirements

sherif566 committed:
Add search functionality to Note and Category controllers

sherif566 committed:
Add Traits to uniform the way responses are returned in each method in Note nad Category Controllers

sherif566 committed:
Add detailed logging to NoteController methods

sherif566 committed:
Include category_id in NoteResource to enhance API responses

sherif566 committed:
Add return types to service and repository methods for improved type safety

sherif566 committed:
Update naming of methods in CategoryNote service and controller

sherif566 committed:
Refactor controllers to use Spatie DTOs for streamlined data handling

sherif566 committed:
Refactor Category and Note repositories to extend BaseRepository

sherif566 committed:
Rename Repository Interface to CRUDInterface as it is more descriptive

sherif566 committed:
Implement centralized pagination constant in application utilities

sherif566 committed:
Rearrange Category model definitions of variables and functions to make sure it adheres to Laravel's conventional structuring

sherif566 committed:
Refactor validation to use model class in CategoryRequest

sherif566 committed:
Remove Note ID from Log

sherif566 committed:
Register the Validation Service in the Service Container and use it in the CategoryNote Controller

sherif566 committed:
Register the Validation Service in the Service Container and use it in the CategoryNote Controller

sherif566 committed:
Refactor CategoryNoteController to delegate business logic to CategoryNoteService

sherif566 committed:
Refactor all controllers to use from() with validated data for DTOs

sherif566 committed:
feat: Add index to the 'description' field in the notes table for optimized search performance

NadaMohamedOptimalSolutions commented on app/DTOs/CategoryDTO.php:
Please review the official package documentation to create the DTO: https://spatie.be/docs/laravel-data/v4/as-a-data-transfer-object/creating-a-data-object.

The problem with the current implementation is that you will need to define the mapping for any new fields. For example, if a new column called description is added to the categories table, this field must be added to the DTO twice—once as a property and another in the from method.

NadaMohamedOptimalSolutions commented on app/Exceptions/Handler.php:
```diff
@@ -24,7 +29,67 @@ class Handler extends ExceptionHandler
+     */
+    protected function invalidJson($request, ValidationException $exception)
+    {
+        return response()->json([
```
Please return the response here using the RespondsWithHttpStatus trait

NadaMohamedOptimalSolutions commented on app/Traits/RespondsWithHttpStatus.php:
```diff
@@ -0,0 +1,30 @@
+
+use Illuminate\Http\JsonResponse;
+
+trait RespondsWithHttpStatus
```
How will you handle the pagination? For example, from the category index response, the categories should be paginated. How will this be handled? The current code does not handle pagination.

NadaMohamedOptimalSolutions commented on app/Traits/RespondsWithHttpStatus.php:
```diff
@@ -0,0 +1,30 @@
+{
+    protected function success($data, $message = '', $status = JsonResponse::HTTP_OK): JsonResponse
+    {
+        if ($data instanceof \Illuminate\Http\Resources\Json\ResourceCollection) {
```
Instead of using an `if` statement, you can create a separate method to handle returning a collection. This adheres to the **Single Responsibility Principle**, making the code cleaner and easier to maintain.

NadaMohamedOptimalSolutions commented on app/Traits/RespondsWithHttpStatus.php:
```diff
@@ -0,0 +1,30 @@
+        ], $status);
+    }
+
+    protected function error($message, $data = [], $status = JsonResponse::HTTP_NOT_FOUND): JsonResponse
```
Please remove the default error code from here, as there might be many different error codes returned, not just `404 Not Found`. For success responses, `200` is sufficient, but unfortunately, there is no unified code for general errors. For example:
- If there is a validation issue, `404 Not Found` would not be appropriate and would be misleading.
- If the problem is server-related, a `5xx` error should be returned instead.

So, it’s better to remove the default value of the error code to ensure the correct code is explicitly provided when calling this method.

NadaMohamedOptimalSolutions commented on config/PaginationConfig.php:
Please move the `PaginationConfig` class to a new directory `App\Constants`. The `Config` directory already exists and refers to Laravel's configuration files. To avoid confusion, it's better to place this class in a `Constants` directory, which is more appropriate for defining static constants like `DEFAULT_PAGE_SIZE`.

NadaMohamedOptimalSolutions commented on database/migrations/2025_01_09_173601_add_index_to_description_in_notes_table.php:
```diff
@@ -0,0 +1,28 @@
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Support\Facades\Schema;
+
+class AddIndexToDescriptionInNotesTable extends Migration
```
Please define the migration as an anonymous class to adhere to Laravel's conventions introduced in Laravel 8. Before this, migration classes were explicitly named (e.g., `class CreateUsersTable`), which could lead to conflicts when multiple migrations had the same class name. Using anonymous classes avoids such conflicts, especially in larger teams or projects.

NadaMohamedOptimalSolutions commented on database/migrations/2025_01_09_173601_add_index_to_description_in_notes_table.php:
```diff
@@ -0,0 +1,28 @@
+    public function up(): void
+    {
+        Schema::table('notes', function (Blueprint $table) {
+            $table->index('description'); // Add an index to the description column
```
The description column in the notes table is of type text. Adding a B-tree index on this column may not be beneficial because indexes on text or varchar columns are not effectively used when performing searches like LIKE '%search_word%'. This is due to how B-tree indexes work, as they cannot optimize searches that begin with a wildcard (%).

I do not recommend adding this index on the description column. Instead, consider which columns are frequently queried or filtered in your application, and determine if they could benefit from an index. For example:

Exact Matches: Indexes are effective for columns that are searched using exact matches (e.g., WHERE column = value).
Prefix Matches: B-tree indexes can also optimize queries like LIKE 'search_word%'.
Please review the query patterns for your application and think about adding effective indexes on columns that would benefit the most. Let me know if you'd like help identifying such columns or implementing proper indexes!

NadaMohamedOptimalSolutions commented on routes/web.php:
```diff
@@ -16,3 +17,8 @@
     return view('welcome');
 });
+
+Route::get('/test-log', function () {
```
Please remove the route `/test-log` as it will not be used in the application.

NadaMohamedOptimalSolutions commented on app/Utilities/Constants.php:
```diff
@@ -0,0 +1,8 @@
+<?php
```
Why do you need this class as well as the App\Config class?

NadaMohamedOptimalSolutions commented on app/Services/ValidationService.php:
Creating a **service** for validation is not the best approach. Instead, you can create a **Validator** class or another dedicated class to handle validation. Services are better suited for handling business logic, while a Validator class focuses solely on validation responsibilities. 

Additionally, in this `validateNoteCategory` method, you can throw an exception if the validation fails. This allows any service using the validator to simply call the method without needing to handle its response, simplifying the logic and improving code readability.

NadaMohamedOptimalSolutions commented on app/Services/NoteService.php:
```diff
@@ -0,0 +1,55 @@
+use App\Models\Note;
+use Illuminate\Database\Eloquent\Collection;
+
+class NoteService
```
The methods create, update, delete are not needed here as they will be handled by the `CategoryNoteService`

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,69 @@
+
+    public function index(Request $request): JsonResponse
+    {
+        $query = Category::query();
```
For the search:

Category:
Should be searchable by parent_id and name (either separately or together).

Note:
Should be searchable by category_id, title (either separately or together). You can also add a search by description if needed.

Please handle getting the search parameters using a request class.

Note:
Since the search will also occur in the index method, both the Category and Note indexes should accept pagination parameters (per_page and page, or whatever you choose to name them). Consider how you will handle both pagination and search parameters together.

Use a DTO to transfer which columns to be searched from the controller to the service.

Define the search logic in the appropriate layer of the application.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        $this->categoryService->delete($category);
+        Log::info('Deleted category', ['category_id' => $category->id]);
+        return $this->success(null, 'Category deleted successfully', JsonResponse::HTTP_OK);
```
You can simply return a message instead of a null object. Please add a new method to handle returning a success message.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,76 @@
+{
+    use RespondsWithHttpStatus;
+
+    private CategoryNoteService $categoryNoteService;
```
Please define this service as other services are defined in the constructor

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,76 @@
+
+    public function store(NoteRequest $request, Category $category): JsonResponse
+    {
+        $dto = NoteDTO::from($request->validated() + ['category_id' => $category->id]);
```
Please utilize the package to get the category_id from the request

NadaMohamedOptimalSolutions commented on app/Http/Controllers/NoteController.php:
```diff
@@ -0,0 +1,68 @@
+        $this->noteService->delete($note);
+        Log::info('Note deleted successfully');
+
+        return $this->success(null, 'Note deleted successfully', 200);
```
Do not return 200 use the JsonResponse status code as you did in other controllers

NadaMohamedOptimalSolutions commented on app/Repositories/BaseRepository.php:
```diff
@@ -0,0 +1,39 @@
+
+    public function create(array $data): Model
+    {
+        return $this->model->create($data);
```
Please return `$this->model->create($data)->fresh()`
You can learn more about fresh and refresh here https://dev.to/bhaidar/the-difference-between-fresh-and-refresh-in-laravel-eloquent-3okj

NadaMohamedOptimalSolutions commented on app/Repositories/Interfaces/CRUDInterface.php:
```diff
@@ -0,0 +1,14 @@
+
+interface CRUDInterface
+{
+    public function all($perPage = 10): LengthAwarePaginator;
```
Please use a constant instead of the number 10

NadaMohamedOptimalSolutions commented on app/Services/CategoryNoteService.php:
```diff
@@ -0,0 +1,43 @@
+
+    public function update(array $data, Category $category, Note $note): Note
+    {
+        if ($note->category_id !== $category->id) {
```
Please use the validation method you defined

NadaMohamedOptimalSolutions commented on app/Services/CategoryNoteService.php:
Please use the note repository to create, update and delete the note.

NadaMohamedOptimalSolutions commented on app/Services/CategoryService.php:
```diff
@@ -0,0 +1,44 @@
+
+    public function create(CategoryDTO $dto): Category
+    {
+        return $this->categoryRepository->create([
```
Please use the `toArray()` method which is defined in the spatie Laravel-data package

NadaMohamedOptimalSolutions commented on app/Models/Category.php:
```diff
@@ -0,0 +1,33 @@
+
+    protected $fillable = ['name', 'parent_id'];
+
+    public static function getTableName()
```
Please remove this unused method

NadaMohamedOptimalSolutions commented on app/Repositories/BaseRepository.php:
Please use a template as a type placeholder (e.g., @template T) to define the general functionality of the CRUD repository. This will ensure type safety and allow the repository to work with specific models while maintaining flexibility and reusability. Using a template will also improve IDE support and make the repository's methods more predictable and consistent.

NadaMohamedOptimalSolutions commented on app/Services/NoteService.php:
```diff
@@ -0,0 +1,55 @@
+        $this->noteRepository = $noteRepository;
+    }
+
+    public function getAll($search = null): Collection
```
Will the method getAll return a collection or a LengthAwarePaginator? Some services return a collection and others return LengthAwarePaginator, please make sure to unify what this methods returns.

NadaMohamedOptimalSolutions commented on app/Repositories/BaseRepository.php:
```diff
@@ -0,0 +1,39 @@
+        $this->model = $model;
+    }
+
+    public function all($perPage = Constants::DEFAULT_PAGINATION): LengthAwarePaginator
```
Please add logs to all the repository methods

NadaMohamedOptimalSolutions commented on app/Repositories/BaseRepository.php:
```diff
@@ -0,0 +1,39 @@
+{
+    protected Model $model;
+
+    public function __construct(Model $model)
```
Please pass the model itself, instead of using and defining `$this->model`
Please make sure to change the other repositories accordingly, and remove any unneeded validation

sherif566 committed:
refactor(DTO):streamline DTO implementation by using reflection-based auto-mapping

sherif566 committed:
refactor(DTO):streamline DTO implementation by using reflection-based auto-mapping

sherif566 committed:
refactor: Implement search and pagination for Categories and Notes

sherif566 committed:
refactor: Implement search and pagination for Categories and Notes

NadaMohamedOptimalSolutions commented on app/DTOs/CategorySearchDTO.php:
```diff
@@ -0,0 +1,13 @@
+
+class CategorySearchDTO extends DataTransferObject
+{
+    public ?string $name = null;        // Nullable: Search by category name
```
1. Please remove the comments as they do not add value.
2. Does per_page and page belong in this class? This class should handle only search arguments, and it does not need to hold any information about pagination. Including pagination here violates the Single Responsibility Principle (SRP).
3. Please move the pagination-related values to a separate constant class, so these numbers can be reused across the application instead of having hardcoded values in the codebase.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,69 @@
+    {
+        $dto = new CategorySearchDTO($request->validated());
+
+        $categories = $this->categoryService->searchCategories($dto);
```
As the search will not always happen it is better to call a more generic method, maybe getCategories, or getAll. I suggest renaming it to getAll to keep the method name consistent among services.

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryController.php:
```diff
@@ -0,0 +1,69 @@
+        return $this->success(new CategoryResource($category), 'Category details retrieved successfully');
+    }
+
+    public function store(CategoryRequest $request): JsonResponse
```
Please add logging in the beginning of the method as well

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,76 @@
+    {
+        $dto = new NoteDTO($request->validated() + ['category_id' => $category->id]);
+
+        $note = $this->categoryNoteService->create($dto->toArray(), $category);
```
Please send the DTO instead of an array. The DTOs are used to pass defined data through the application layers instead of using an array which we are not sure about its structure

NadaMohamedOptimalSolutions commented on app/Http/Requests/CategorySearchRequest.php:
```diff
@@ -0,0 +1,31 @@
+        return [
+            'name' => ['nullable', 'string', 'max:255'],
+            'parent_id' => ['nullable', 'integer'],
+            'per_page' => ['nullable', 'integer', 'min:1'],
```
Please move the repeated per_page and page validation rules into a separate request class. This will ensure the code adheres to the DRY (Don't Repeat Yourself) principle and follows the Single Responsibility Principle (SRP) by keeping pagination concerns separate from specific request validations.

sherif566 committed:
refactor: Extend DTOs from Data instead of DataTransferObject

sherif566 committed:
refactor: adjust DTOS to follow spatie doscumentation

sherif566 committed:
refactor: Adjusted controller to utilize From keyword to make use of the automatic mapping by spatie

sherif566 committed:
refactor: Note $ Category service layers are adjusted to use toarray() method instead of manually passing properties of the DTO

sherif566 committed:
refactor(functions): rename searchnotes & searchcategories to search only because it is already clear

sherif566 committed:
Refactor: Extract pagination rules into a separate request class

sherif566 committed:
refactor: adjusted the loggings in all controllers to be consistent

sherif566 committed:
refactor: the dependency injection in the category service

sherif566 committed:
feat(dto): refactor CategorySearchDTO to extend PaginationDTO

sherif566 committed:
refactor:removed the search from base repository and implemented it separately in each repository

sherif566 committed:
refactor: separate pagination logic and simplify service search function

- Moved pagination logic to a dedicated function in repositories.
- Updated CategoryService::search to pass CategorySearchDTO directly to the repository.
- Removed manual extraction of per_page and page, utilizing PaginationDTO instead.
- Improved code maintainability and separation of concerns.

NadaMohamedOptimalSolutions commented on app/DTOs/NoteSearchDTO.php:
```diff
@@ -0,0 +1,17 @@
+    public ?string $title = null,
+    public ?string $description = null,
+    public ?int $category_id = null,
+    public PaginationDTO $pagination
```
1. You do not need to add the property public PaginationDTO $pagination because the class already extends PaginationDTO, meaning it inherits its properties and methods. 
2. Since PaginationDTO has a constructor with default values (per_page = 15 and page = 1), calling parent::__construct(); in NoteSearchDTO is not required

NadaMohamedOptimalSolutions commented on app/Http/Controllers/CategoryNoteController.php:
```diff
@@ -0,0 +1,93 @@
+    {
+        Log::info('Fetching notes for category started.', ['category_id' => $category->id]);
+
+        $notes = $this->categoryNoteService->getAll($category);
```
I suggest naming all the methods called in the index **`getAll`** instead of **`search`**. The reason is that the **index method does not always perform a search**—it might simply retrieve all records with constraints, rather than executing a full search operation. The term **"search"** implies an active filtering process, while **"getAll"** aligns better with both unrestricted and constrained retrieval.

NadaMohamedOptimalSolutions commented on app/Http/Requests/CategorySearchRequest.php:
```diff
@@ -0,0 +1,17 @@
+    {
+        return array_merge(parent::rules(), [
+            'name' => ['nullable', 'string', 'max:255'],
+            'parent_id' => ['nullable', 'integer'],
```
Please add a validation rule to ensure that the provided parent_id exists in the categories table. This prevents referencing a non-existent category.

NadaMohamedOptimalSolutions commented on app/Http/Requests/NoteSearchRequest.php:
```diff
@@ -0,0 +1,18 @@
+        return array_merge(parent::rules(), [
+            'title' => ['nullable', 'string', 'max:255'],
+            'description' => ['nullable', 'string'],
+            'category_id' => ['nullable', 'integer'],
```
Please add a validation rule to ensure that the provided category_id exists in the categories table. This prevents referencing a non-existent category.

sherif566 committed:
feat: refactor dependency injection for improved code clarity and maintainability

sherif566 committed:
feat(logging): Add logging to controllers, services, and repositories
Implemented logging at the beginning and end of each function across all project layers.
Added contextual information (IDs, request data, operation status) for better traceability.
Used info logs in controllers and services, and debug logs in repositories for detailed insights.

sherif566 committed:
refactor: Update index methods to call getAll, integrating search and retrieval

Modified index functions in CategoryController and NoteController to call getAll in their respective services.
Updated getAll in CategoryService and NoteService to handle both search and retrieval of all records.

NadaMohamedOptimalSolutions commented on app/Repositories/CategoryRepository.php:
```diff
@@ -0,0 +1,56 @@
+
+        Log::debug('Category search query executed successfully');
+
+        return $this->paginate($query);
```
You'll need to pass the pagination arguments as well to the paginate method.

NadaMohamedOptimalSolutions commented on app/Services/CategoryService.php:
```diff
@@ -0,0 +1,82 @@
+
+
+
+    public function search(CategorySearchDTO $dto): LengthAwarePaginator
```
Is this method used?
